File: apps/api-rs/src/llm/openai.rs
use std::time::Duration;

use anyhow::{anyhow, Result};
use async_trait::async_trait;
use reqwest::StatusCode;
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone)]
pub struct OpenAiClientConfig {
    pub base_url: String,
    pub api_key: String,
    pub model: String,
    pub timeout_seconds: u64,
}

impl Default for OpenAiClientConfig {
    fn default() -> Self {
        Self {
            base_url: "http://localhost:11434/v1".to_string(),
            api_key: "ollama".to_string(),
            model: "qwen2.5:14b".to_string(),
            timeout_seconds: 120,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct ChatMessage {
    pub role: String,
    pub content: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct ChatCompletionRequest {
    pub model: String,
    pub messages: Vec<ChatMessage>,
    pub temperature: f64,
    pub max_tokens: u32,
    pub stream: bool,
}

#[derive(Debug, Clone, Deserialize)]
pub struct DeltaChoice {
    #[serde(default)]
    pub index: i32,
    pub delta: Option<Value>,
    pub finish_reason: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct StreamChunk {
    pub id: Option<String>,
    pub object: Option<String>,
    pub choices: Option<Vec<DeltaChoice>>,
}

#[derive(Debug, Clone)]
pub struct LlmStreamError {
    pub code: String,
    pub message: String,
}

impl LlmStreamError {
    pub fn provider(status: StatusCode, message: String) -> Self {
        let code = if status == StatusCode::UNAUTHORIZED {
            "LLM_UNAUTHORIZED"
        } else {
            "LLM_PROVIDER_ERROR"
        };
        Self {
            code: code.to_string(),
            message,
        }
    }

    pub fn stream(message: impl Into<String>) -> Self {
        Self {
            code: "LLM_STREAM_ERROR".to_string(),
            message: message.into(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum ParsedStreamEvent {
    Delta(String),
    Done,
    Error(String),
    Empty,
}

#[async_trait]
pub trait LlmClient: Send + Sync {
    async fn complete_json<T>(&self, prompt: String, system: Option<String>) -> Result<T>
    where
        T: serde::de::DeserializeOwned;

    async fn complete_json_with_options<T>(
        &self,
        prompt: String,
        system: Option<String>,
        temperature: f64,
        max_tokens: u32,
    ) -> Result<T>
    where
        T: serde::de::DeserializeOwned;

    async fn stream_text(
        &self,
        prompt: String,
        system: Option<String>,
        temperature: f64,
        max_tokens: u32,
    ) -> Result<tokio::sync::mpsc::UnboundedReceiver<Result<String, LlmStreamError>>>;

    async fn stream_chat(
        &self,
        messages: Vec<ChatMessage>,
        temperature: f64,
        max_tokens: u32,
    ) -> Result<tokio::sync::mpsc::UnboundedReceiver<Result<String, LlmStreamError>>>;
}

pub struct OpenAiClient {
    pub(super) config: OpenAiClientConfig,
    pub(super) http: reqwest::Client,
}

impl OpenAiClient {
    pub fn new(config: OpenAiClientConfig) -> Result<Self> {
        let http = reqwest::Client::builder()
            .timeout(Duration::from_secs(config.timeout_seconds))
            .build()?;
        Ok(Self { config, http })
    }

    pub(super) fn chat_url(&self) -> String {
        format!(
            "{}/chat/completions",
            self.config.base_url.trim_end_matches('/')
        )
    }

    pub(super) fn auth_header(&self) -> String {
        format!("Bearer {}", self.config.api_key)
    }

    async fn request_json<T>(
        &self,
        prompt: String,
        system: Option<String>,
        temperature: f64,
        max_tokens: u32,
    ) -> Result<T>
    where
        T: serde::de::DeserializeOwned,
    {
        let mut messages = vec![];
        if let Some(system) = system {
            messages.push(ChatMessage {
                role: "system".to_string(),
                content: system,
            });
        }
        messages.push(ChatMessage {
            role: "user".to_string(),
            content: prompt,
        });
        let request = ChatCompletionRequest {
            model: self.config.model.clone(),
            messages,
            temperature,
            max_tokens,
            stream: false,
        };
        let response = self
            .http
            .post(self.chat_url())
            .header("Authorization", self.auth_header())
            .header("Content-Type", "application/json")
            .json(&request)
            .send()
            .await?
            .error_for_status()?;
        let payload: Value = response.json().await?;
        let content = payload["choices"][0]["message"]["content"]
            .as_str()
            .ok_or_else(|| anyhow!("missing content in completion response"))?;
        parse_json_completion(content)
    }
}

#[async_trait]
impl LlmClient for OpenAiClient {
    async fn complete_json<T>(&self, prompt: String, system: Option<String>) -> Result<T>
    where
        T: serde::de::DeserializeOwned,
    {
        self.request_json(prompt, system, 0.2, 2048).await
    }

    async fn complete_json_with_options<T>(
        &self,
        prompt: String,
        system: Option<String>,
        temperature: f64,
        max_tokens: u32,
    ) -> Result<T>
    where
        T: serde::de::DeserializeOwned,
    {
        self.request_json(prompt, system, temperature, max_tokens)
            .await
    }

    async fn stream_text(
        &self,
        prompt: String,
        system: Option<String>,
        temperature: f64,
        max_tokens: u32,
    ) -> Result<tokio::sync::mpsc::UnboundedReceiver<Result<String, LlmStreamError>>> {
        let mut messages = vec![];
        if let Some(s) = system {
            messages.push(ChatMessage {
                role: "system".to_string(),
                content: s,
            });
        }
        messages.push(ChatMessage {
            role: "user".to_string(),
            content: prompt,
        });

        self.stream_chat(messages, temperature, max_tokens).await
    }

    async fn stream_chat(
        &self,
        messages: Vec<ChatMessage>,
        temperature: f64,
        max_tokens: u32,
    ) -> Result<tokio::sync::mpsc::UnboundedReceiver<Result<String, LlmStreamError>>> {
        let req = ChatCompletionRequest {
            model: self.config.model.clone(),
            messages,
            temperature,
            max_tokens,
            stream: true,
        };

        let resp = self
            .http
            .post(self.chat_url())
            .header("Authorization", self.auth_header())
            .header("Content-Type", "application/json")
            .json(&req)
            .send()
            .await?;

        let status = resp.status();
        if !status.is_success() {
            let body = resp
                .text()
                .await
                .unwrap_or_else(|_| "LLM provider returned an unreadable error body".to_string());
            let message = provider_error_message(status, &body);
            let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
            let _ = tx.send(Err(LlmStreamError::provider(status, message)));
            return Ok(rx);
        }

        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
        let mut stream = resp.bytes_stream();

        tokio::spawn(async move {
            let mut buffer = String::new();
            while let Some(chunk) = stream.next().await {
                match chunk {
                    Ok(bytes) => {
                        buffer.push_str(&String::from_utf8_lossy(&bytes));
                        while let Some((pos, sep_len)) = find_sse_separator(&buffer) {
                            let event_text = buffer[..pos].to_string();
                            buffer = buffer[pos + sep_len..].to_string();
                            match parse_sse_event(&event_text) {
                                ParsedStreamEvent::Delta(text) => {
                                    let _ = tx.send(Ok(text));
                                }
                                ParsedStreamEvent::Error(message) => {
                                    let _ = tx.send(Err(LlmStreamError::stream(message)));
                                    return;
                                }
                                ParsedStreamEvent::Done => return,
                                ParsedStreamEvent::Empty => {}
                            }
                        }
                    }
                    Err(e) => {
                        tracing::error!("llm stream error: {e}");
                        let _ = tx.send(Err(LlmStreamError::stream(e.to_string())));
                        break;
                    }
                }
            }

            let rest = buffer.trim();
            if !rest.is_empty() {
                match parse_sse_event(rest) {
                    ParsedStreamEvent::Delta(text) => {
                        let _ = tx.send(Ok(text));
                    }
                    ParsedStreamEvent::Error(message) => {
                        let _ = tx.send(Err(LlmStreamError::stream(message)));
                    }
                    ParsedStreamEvent::Done | ParsedStreamEvent::Empty => {}
                }
            }
        });

        Ok(rx)
    }
}

fn find_sse_separator(buffer: &str) -> Option<(usize, usize)> {
    match (buffer.find("\n\n"), buffer.find("\r\n\r\n")) {
        (Some(a), Some(b)) if a < b => Some((a, 2)),
        (Some(_), Some(b)) => Some((b, 4)),
        (Some(a), None) => Some((a, 2)),
        (None, Some(b)) => Some((b, 4)),
        (None, None) => None,
    }
}

fn parse_sse_event(event_text: &str) -> ParsedStreamEvent {
    let mut data = String::new();
    for line in event_text.lines() {
        let trimmed = line.trim();
        if trimmed.starts_with("data:") {
            let payload = trimmed.trim_start_matches("data:").trim();
            if payload == "[DONE]" {
                return ParsedStreamEvent::Done;
            }
            if !data.is_empty() {
                data.push('\n');
            }
            data.push_str(payload);
        }
    }
    if data.is_empty() {
        return ParsedStreamEvent::Empty;
    }
    let value: Value = match serde_json::from_str(&data) {
        Ok(value) => value,
        Err(_) => return ParsedStreamEvent::Empty,
    };
    if let Some(message) = value
        .get("error")
        .and_then(|error| error.get("message"))
        .and_then(Value::as_str)
    {
        return ParsedStreamEvent::Error(message.to_string());
    }
    let chunk: StreamChunk = match serde_json::from_value(value) {
        Ok(chunk) => chunk,
        Err(_) => return ParsedStreamEvent::Empty,
    };
    let Some(choices) = chunk.choices else {
        return ParsedStreamEvent::Empty;
    };
    let Some(choice) = choices.first() else {
        return ParsedStreamEvent::Empty;
    };
    if choice.finish_reason.is_some() {
        return ParsedStreamEvent::Done;
    }
    let Some(delta) = choice.delta.as_ref() else {
        return ParsedStreamEvent::Empty;
    };
    delta["content"]
        .as_str()
        .filter(|s| !s.is_empty())
        .map(|s| ParsedStreamEvent::Delta(s.to_string()))
        .unwrap_or(ParsedStreamEvent::Empty)
}

fn provider_error_message(status: StatusCode, body: &str) -> String {
    if let Ok(value) = serde_json::from_str::<Value>(body) {
        if let Some(message) = value
            .get("error")
            .and_then(|error| error.get("message"))
            .and_then(Value::as_str)
        {
            return format!("LLM provider returned {status}: {message}");
        }
        if let Some(message) = value.get("message").and_then(Value::as_str) {
            return format!("LLM provider returned {status}: {message}");
        }
    }
    let trimmed = body.trim();
    if trimmed.is_empty() {
        format!("LLM provider returned {status}")
    } else {
        format!("LLM provider returned {status}: {trimmed}")
    }
}

fn strip_json_fences(text: &str) -> String {
    let trimmed = text.trim();
    if trimmed.starts_with("```json") && trimmed.ends_with("```") {
        let inner = trimmed
            .trim_start_matches("```json")
            .trim_end_matches("```");
        return inner.trim().to_string();
    }
    if trimmed.starts_with("```") && trimmed.ends_with("```") {
        let inner = trimmed.trim_start_matches("```").trim_end_matches("```");
        return inner.trim().to_string();
    }
    trimmed.to_string()
}

fn parse_json_completion<T>(text: &str) -> Result<T>
where
    T: serde::de::DeserializeOwned,
{
    let stripped = strip_json_fences(text);
    match serde_json::from_str(&stripped) {
        Ok(value) => Ok(value),
        Err(original_error) => {
            let repaired = escape_control_characters_in_json_strings(&stripped);
            serde_json::from_str(&repaired).map_err(|repaired_error| {
                anyhow!(
                    "invalid JSON completion: {original_error}; control-character repair also failed: {repaired_error}"
                )
            })
        }
    }
}

fn escape_control_characters_in_json_strings(text: &str) -> String {
    let mut output = String::with_capacity(text.len());
    let mut in_string = false;
    let mut escaped = false;
    for character in text.chars() {
        if !in_string {
            output.push(character);
            if character == '"' {
                in_string = true;
            }
            continue;
        }
        if escaped {
            output.push(character);
            escaped = false;
            continue;
        }
        match character {
            '\\' => {
                output.push(character);
                escaped = true;
            }
            '"' => {
                output.push(character);
                in_string = false;
            }
            '\n' => output.push_str("\\n"),
            '\r' => output.push_str("\\r"),
            '\t' => output.push_str("\\t"),
            '\u{0008}' => output.push_str("\\b"),
            '\u{000C}' => output.push_str("\\f"),
            control if control <= '\u{001F}' => {
                output.push_str(&format!("\\u{:04X}", control as u32));
            }
            _ => output.push(character),
        }
    }
    output
}

use futures::StreamExt;

#[cfg(test)]
#[path = "openai_tests.rs"]
mod tests;

File: apps/api-bun/src/llm/openai.ts
// 移植自 apps/api-rs/src/llm/openai.rs —— OpenAI 兼容客户端
import type { Usage } from '../models/index.ts';

export interface OpenAiClientConfig {
  baseUrl: string; apiKey: string; model: string; timeoutSeconds: number;
}
export function defaultOpenAiClientConfig(): OpenAiClientConfig {
  return { baseUrl: 'http://localhost:11434/v1', apiKey: 'ollama', model: 'qwen2.5:14b', timeoutSeconds: 120 };
}

export interface ChatMessage { role: string; content: string; }

export interface LlmStreamErrorShape { code: string; message: string; }
export function llmStreamProviderError(status: number, message: string): LlmStreamErrorShape {
  return { code: status === 401 ? 'LLM_UNAUTHORIZED' : 'LLM_PROVIDER_ERROR', message };
}
export function llmStreamError(message: string): LlmStreamErrorShape {
  return { code: 'LLM_STREAM_ERROR', message };
}

export type ParsedStreamEvent =
  | { kind: 'delta'; text: string }
  | { kind: 'done' }
  | { kind: 'error'; message: string }
  | { kind: 'empty' };

export function findSseSeparator(buffer: string): [number, number] | null {
  const unix = buffer.indexOf('\n\n');
  const windows = buffer.indexOf('\r\n\r\n');
  if (unix !== -1 && windows !== -1) return unix < windows ? [unix, 2] : [windows, 4];
  if (unix !== -1) return [unix, 2];
  if (windows !== -1) return [windows, 4];
  return null;
}

export function parseSseEvent(eventText: string): ParsedStreamEvent {
  let data = '';
  for (const line of eventText.split(/\r?\n/)) {
    const trimmed = line.trim();
    if (trimmed.startsWith('data:')) {
      const payload = trimmed.slice('data:'.length).trim();
      if (payload === '[DONE]') return { kind: 'done' };
      if (data.length > 0) data += '\n';
      data += payload;
    }
  }
  if (data.length === 0) return { kind: 'empty' };
  let value: unknown;
  try { value = JSON.parse(data); } catch { return { kind: 'empty' }; }
  if (typeof value !== 'object' || value === null) return { kind: 'empty' };
  const record = value as Record<string, unknown>;
  const error = record.error;
  if (typeof error === 'object' && error !== null) {
    const message = (error as Record<string, unknown>).message;
    if (typeof message === 'string') return { kind: 'error', message };
  }
  const choices = record.choices;
  if (!Array.isArray(choices) || choices.length === 0) return { kind: 'empty' };
  const choice = choices[0] as Record<string, unknown>;
  if (choice.finish_reason != null) return { kind: 'done' };
  const delta = choice.delta;
  if (typeof delta !== 'object' || delta === null) return { kind: 'empty' };
  const content = (delta as Record<string, unknown>).content;
  if (typeof content === 'string' && content.length > 0) return { kind: 'delta', text: content };
  return { kind: 'empty' };
}

export function providerErrorMessage(status: number, body: string): string {
  try {
    const value = JSON.parse(body) as Record<string, unknown>;
    const error = value.error;
    if (typeof error === 'object' && error !== null) {
      const message = (error as Record<string, unknown>).message;
      if (typeof message === 'string') return `LLM provider returned ${status}: ${message}`;
    }
    if (typeof value.message === 'string') return `LLM provider returned ${status}: ${value.message}`;
  } catch { /* fall through */ }
  const trimmed = body.trim();
  return trimmed.length === 0
    ? `LLM provider returned ${status}`
    : `LLM provider returned ${status}: ${trimmed}`;
}

export function stripJsonFences(text: string): string {
  const trimmed = text.trim();
  if (trimmed.startsWith('```json') && trimmed.endsWith('```')) {
    return trimmed.slice('```json'.length, trimmed.length - 3).trim();
  }
  if (trimmed.startsWith('```') && trimmed.endsWith('```')) {
    return trimmed.slice(3, trimmed.length - 3).trim();
  }
  return trimmed;
}

export function escapeControlCharactersInJsonStrings(text: string): string {
  let output = '';
  let inString = false;
  let escaped = false;
  for (const character of text) {
    if (!inString) {
      output += character;
      if (character === '"') inString = true;
      continue;
    }
    if (escaped) { output += character; escaped = false; continue; }
    switch (character) {
      case '\\': output += character; escaped = true; break;
      case '"': output += character; inString = false; break;
      case '\n': output += '\\n'; break;
      case '\r': output += '\\r'; break;
      case '\t': output += '\\t'; break;
      case '\b': output += '\\b'; break;
      case '\f': output += '\\f'; break;
      default:
        if (character <= '\u001f') {
          output += '\\u' + character.charCodeAt(0).toString(16).toUpperCase().padStart(4, '0');
        } else {
          output += character;
        }
    }
  }
  return output;
}

export function parseJsonCompletion<T>(text: string): T {
  const stripped = stripJsonFences(text);
  try {
    return JSON.parse(stripped) as T;
  } catch (originalError) {
    const repaired = escapeControlCharactersInJsonStrings(stripped);
    try {
      return JSON.parse(repaired) as T;
    } catch (repairedError) {
      throw new Error(
        `invalid JSON completion: ${(originalError as Error).message}; control-character repair also failed: ${(repairedError as Error).message}`);
    }
  }
}

interface ChatCompletionPayload {
  model: string; messages: ChatMessage[]; temperature: number; max_tokens: number; stream: boolean;
}

export class OpenAiClient {
  readonly config: OpenAiClientConfig;

  constructor(config: OpenAiClientConfig) {
    this.config = config;
  }

  chatUrl(): string {
    return `${this.config.baseUrl.replace(/\/+$/, '')}/chat/completions`;
  }
  authHeader(): string { return `Bearer ${this.config.apiKey}`; }

  private async requestJson<T>(
    prompt: string, system: string | null, temperature: number, maxTokens: number,
  ): Promise<T> {
    const messages: ChatMessage[] = [];
    if (system !== null) messages.push({ role: 'system', content: system });
    messages.push({ role: 'user', content: prompt });
    const payload: ChatCompletionPayload = {
      model: this.config.model, messages, temperature, max_tokens: maxTokens, stream: false,
    };
    const response = await fetch(this.chatUrl(), {
      method: 'POST',
      headers: { Authorization: this.authHeader(), 'Content-Type': 'application/json' },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(this.config.timeoutSeconds * 1000),
    });
    if (!response.ok) {
      throw new Error(providerErrorMessage(response.status, await response.text()));
    }
    const payloadJson = (await response.json()) as Record<string, unknown>;
    const choices = (payloadJson.choices ?? []) as Array<Record<string, unknown>>;
    const firstChoice = choices[0];
    const content = firstChoice === undefined
      ? undefined
      : (firstChoice.message as Record<string, unknown> | undefined)?.content;
    if (typeof content !== 'string') throw new Error('missing content in completion response');
    return parseJsonCompletion<T>(content);
  }

  completeJson<T>(prompt: string, system: string | null): Promise<T> {
    return this.requestJson<T>(prompt, system, 0.2, 2048);
  }
  completeJsonWithOptions<T>(
    prompt: string, system: string | null, temperature: number, maxTokens: number,
  ): Promise<T> {
    return this.requestJson<T>(prompt, system, temperature, maxTokens);
  }

  /** 流式输出：返回异步迭代器，逐条产出 Ok(text) 或 Err(LlmStreamErrorShape)。 */
  async *streamChat(
    messages: ChatMessage[], temperature: number, maxTokens: number,
  ): AsyncGenerator<{ ok: true; text: string } | { ok: false; error: LlmStreamErrorShape }> {
    const payload: ChatCompletionPayload = {
      model: this.config.model, messages, temperature, max_tokens: maxTokens, stream: true,
    };
    let response: Response;
    try {
      response = await fetch(this.chatUrl(), {
        method: 'POST',
        headers: { Authorization: this.authHeader(), 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(this.config.timeoutSeconds * 1000),
      });
    } catch (error) {
      yield { ok: false, error: llmStreamError((error as Error).message) };
      return;
    }
    if (!response.ok || !response.body) {
      const body = await response.text().catch(() => 'LLM provider returned an unreadable error body');
      yield { ok: false, error: llmStreamProviderError(response.status, providerErrorMessage(response.status, body)) };
      return;
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    try {
      for (;;) {
        const { done, value } = await reader.read();
        if (done) break;
        buffer += decoder.decode(value, { stream: true });
        for (;;) {
          const separator = findSseSeparator(buffer);
          if (!separator) break;
          const [position, separatorLength] = separator;
          const eventText = buffer.slice(0, position);
          buffer = buffer.slice(position + separatorLength);
          const event = parseSseEvent(eventText);
          if (event.kind === 'delta') { yield { ok: true, text: event.text }; continue; }
          if (event.kind === 'error') { yield { ok: false, error: llmStreamError(event.message) }; return; }
          if (event.kind === 'done') return;
        }
      }
      const rest = buffer.trim();
      if (rest.length > 0) {
        const event = parseSseEvent(rest);
        if (event.kind === 'delta') yield { ok: true, text: event.text };
        else if (event.kind === 'error') yield { ok: false, error: llmStreamError(event.message) };
      }
    } catch (error) {
      console.error('[documind][llm] stream error:', error);
      yield { ok: false, error: llmStreamError((error as Error).message) };
    } finally {
      reader.cancel().catch(() => undefined);
    }
  }

  streamText(
    prompt: string, system: string | null, temperature: number, maxTokens: number,
  ): AsyncGenerator<{ ok: true; text: string } | { ok: false; error: LlmStreamErrorShape }> {
    const messages: ChatMessage[] = [];
    if (system !== null) messages.push({ role: 'system', content: system });
    messages.push({ role: 'user', content: prompt });
    return this.streamChat(messages, temperature, maxTokens);
  }
}
